src/ray/raylet/scheduling/cluster_resource_scheduler.cc

```diff
@@ -311,9 +311,9 @@ scheduling::NodeID ClusterResourceScheduler::GetBestSchedulableNode(
       !IsSchedulableOnNode(best_node,
                            task_spec.GetRequiredPlacementResources().GetResourceMap(),
                            requires_object_store_memory)) {
-    // Prefer waiting on the local node since the local node is chosen for a reason (e.g.
-    // spread).
-    if (preferred_node_id == local_node_id_.Binary()) {
+    // Prefer waiting on the local node if possible
+    // since the local node is chosen for a reason (e.g. spread).
+    if ((preferred_node_id == local_node_id_.Binary()) && NodeAvailable(local_node_id_)) {
```

Comment:
Q: how can the draining node be the best node from ClusterResourceScheduler, since it should be excluded at https://github.com/ray-project/ray/blob/c18686a39dbffd31868336c59d7274de08ef440b/src/ray/raylet/scheduling/cluster_resource_scheduler.cc#L107 ?

Reply:
My reading call stack:

- GetBestSchedulableNode
- scheduling_policy_->Schedule
- BundlePackSchedulingPolicy::Schedule
- BundleSchedulingPolicy::SelectCandidateNodes
- is_node_available_
- ClusterResourceScheduler::NodeAvailable
- IsNodeDraining

Reply:
I did a quick look at all ISchedulingPolicy impls and it seems all of them called is_node_available_ or is_node_alive_ somewhere, so this is still a mystery